Add count_include_pad arg for average_pool_op on CPU

xiaomengy opened the pull request:
Summary: Add count_include_pad arg for average_pool_op on CPU

Differential Revision: D13558123

houseroad commented on caffe2/operators/pool_op.cc:
```diff
@@ -900,22 +536,20 @@ std::function<void(OpSchema&)> MaxPoolDocGenerator(const char* dim) {
-    PoolOp<float, CPUContext, AveragePool<float>>);
+    PoolOp<float, CPUContext, AveragePoolFunctor<CPUContext>>);
 
 OPERATOR_SCHEMA(AveragePool)
```
Shall we also add the description about arg `count_include_pad` to the AveragePoolDocGenerator?

BIT-silence replied:
Actually I separate this approach into two parts. This is the first part to add it into CPU. And then I will add it into GPU part, which is more important. So I'd suggest add the doc when the GPU part finished.

houseroad commented on caffe2/operators/pool_op_util.cc:
```diff
@@ -0,0 +1,922 @@
+
+#define CAFFE2_SPECIALIZED_RUN_AVERAGE_POOL_1D(T, kCountIncludePad)    \
+  template <>                                                          \
+  void RunAveragePool1D<T, StorageOrder::NCHW, kCountIncludePad>(      \
```
Can we add kCountIncludePad as argument?

BIT-silence replied:
It is trying to make the design of the related functions on CPU and GPU the same. On GPU part, if there is a condition between different threads in runtime, there will be some negative impact on the performance.

houseroad reviewed (commented):
Could you also add a test case to cover `count_include_pad` argument?

houseroad reviewed (approved):
Discussed offline, the comments will be addressed in the following PRs.

xiaomengy committed:
Add count_include_pad arg for average_pool_op on CPU (#15593)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15593

Add count_include_pad arg for average_pool_op on CPU

Reviewed By: houseroad

Differential Revision: D13558123

fbshipit-source-id: 4f29421d760984c639e0e3b9a253c43db855f240
